src/CodeGenerator/src/Generator/RequestSerializer/JsonSerializer.php

```diff
@@ -0,0 +1,303 @@
+<?php
+
+declare(strict_types=1);
+
+namespace AsyncAws\CodeGenerator\Generator\RequestSerializer;
+
+use AsyncAws\CodeGenerator\Definition\ListShape;
+use AsyncAws\CodeGenerator\Definition\MapShape;
+use AsyncAws\CodeGenerator\Definition\Member;
+use AsyncAws\CodeGenerator\Definition\Operation;
+use AsyncAws\CodeGenerator\Definition\Shape;
+use AsyncAws\CodeGenerator\Definition\ShapeWithMembers;
+use AsyncAws\CodeGenerator\Definition\StructureMember;
+use AsyncAws\CodeGenerator\Definition\StructureShape;
+use AsyncAws\CodeGenerator\Generator\Naming\NamespaceRegistry;
+
+/**
+ * Serialize a request body to a nice nested array.
+ *
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ *
+ * @internal
+ */
+class JsonSerializer implements Serializer
+{
+    private $recursionName = null;
+
+    private $recursionInitialized = false;
```

Comment:
Is there a way to avoid state?

What's about adding `requestBody()` in all objects (and not only in Root Input object), Like we do in `Validate`.
That would reduce complexity of the method in root object, and easily fix the recursion thing.

```php
public function requestBody()
{
  $payload = [];
  $payload['LocalProperty'] = $this->LocalProperty;
  if (null !== $this->ObjectProperty) {
     $payload['ObjectProperty'] = $this->ObjectProperty->requestBody();
  }
}
```

Reply:
I thought about that. But I decided not to when I considered xml requests. It will be a bit more complicated but it may work.

Reply:
Your right, it could be done in another PR.

Xml should be easy:
```php
$doc = $document->createElement('Foo');
$doc->appendChild($this->Bar->requestBody()));
```

But, for Query Serializer that could be hard (for the record, the return array is flat: `$payload['Foo.Bar.0.member']`)

Reply:
should be fixed by #318

Reply:
Thank you